src/main/java/org/opensearch/cluster/etcd/changeapplier/DataNodeState.java

```diff
@@ -140,15 +215,59 @@ public ClusterState buildClusterState(ClusterState previousState) {
                 ShardRouting shardRouting;
                 if (previouslyStartedShard.isPresent()) {
                     shardRouting = previouslyStartedShard.get();
+                    logger.debug(
+                        "Reusing existing ShardRouting for shard {}[{}] with allocation ID {}",
+                        indexMetadata.getIndex().getName(),
+                        shardNum,
+                        shardRouting.allocationId().getId()
+                    );
                 } else {
+                    // No previous shard in cluster state - use ETCD-based allocation ID preservation
+                    RecoverySource etcdRecoverySource = determineRecoverySource(shardNum, indexMetadata, role);
+                    String previousAllocationId = getPreviousAllocationId(indexMetadata.getIndex().getName(), shardNum);
```

Comment:
I think these can be considerably simplified now that we're iterating over the `DataNodeShard` instances.